ghcide/src/Development/IDE/Core/Shake.hs

```diff
@@ -1234,7 +1252,9 @@ defineEarlyCutoff' doDiagnostics cmp key file mbOld mode action = do
                     (if eq then ChangedRecomputeSame else ChangedRecomputeDiff)
                     (encodeShakeValue bs) $
                     A res
-        liftIO $ atomicallyNamed "define - dirtyKeys" $ modifyTVar' dirtyKeys (deleteKeySet $ toKey key file)
+        liftIO $ atomicallyNamed "define - (runningKeys, dirtyKeys)" $ do
+            running <- readTVar runningKeys
+            when (memberKeySet theKey running) $ return (deleteKeySet theKey running) >> modifyTVar' dirtyKeys (deleteKeySet theKey)
```

Comment:
This looks suspicious.
deleteKeySet is pure function which creates a new set with the key deleted.
But you seem to be throwing the that new set away (using >>).
I'd expect something like
```haskell
when (memberKeySet theKey running) $ do 
   modifyTVar' runningKeys (deleteKeySet theKey)
   modifyTVar' dirtyKeys (deleteKeySet theKey)
```

Reply:
You are right, I make a mistake, fixed now